File: server/src/agent/ollama.js
const axios = require('axios');
const { managerTools, workerTools, executeWorkerTool } = require('./tools');
const { agentLogger } = require('../utils/logger');
const config = require('../config');

const OLLAMA_BASE_URL = config.ollama.baseUrl;
const OLLAMA_API_KEY = config.ollama.apiKey;

const getHeaders = () => {
    const headers = {};
    if (OLLAMA_API_KEY) {
        headers['Authorization'] = `Bearer ${OLLAMA_API_KEY}`;
    }
    return headers;
};

// ----------------------------------------------------------------------
// WORKER LOOP (Research Agent)
// ----------------------------------------------------------------------

async function runWorkerAgent(query, model, signal, num_ctx, parentOnChunk) {
    agentLogger.info(`[Worker] Starting research on: ${query}`);

    // Worker starts with a fresh context
    const messages = [
        { role: 'system', content: config.ollama.systemMessage.worker },
        { role: 'user', content: query }
    ];

    // Pass null for onChunk to suppress worker token streaming
    return chatLoop(messages, model, signal, num_ctx, workerTools, async (toolName, toolArgs) => {
        // executeWorkerTool wrapper
        if (parentOnChunk) parentOnChunk({ type: 'tool_start', tool: `Worker: ${toolName}` });
        const result = await executeWorkerTool(toolName, toolArgs);
        if (parentOnChunk) parentOnChunk({ type: 'tool_end', tool: `Worker: ${toolName}`, result: 'Done' });
        return result;
    }, null);
}


// ----------------------------------------------------------------------
// MANAGER LOOP (Chat Interface)
// ----------------------------------------------------------------------

async function processUserRequest(messages, model, onChunk, signal, num_ctx) {
    // Ensure system prompt is set for Manager
    const systemMessage = {
        role: 'system',
        content: config.ollama.systemMessage.manager
    };

    let finalMessages = [...messages];
    if (finalMessages.length === 0 || finalMessages[0].role !== 'system') {
        finalMessages = [systemMessage, ...finalMessages];
    }

    // Pass onChunk to allow Manager to stream tokens
    return chatLoop(finalMessages, model, signal, num_ctx, managerTools, async (toolName, toolArgs) => {
        // Manager Tools Execution
        if (toolName === 'delegate_research') {
            if (onChunk) onChunk({ type: 'tool_start', tool: 'Research Agent' });

            // Call the Worker!
            const researchResult = await runWorkerAgent(toolArgs.query, model, signal, num_ctx, onChunk);

            if (onChunk) onChunk({ type: 'tool_end', tool: 'Research Agent', result: 'Research Complete' });

            return `[Research Agent Result]\n${researchResult.content}`;
        }
        return `Error: Unknown manager tool ${toolName}`;
    }, onChunk);
}


// ----------------------------------------------------------------------
// GENERIC CHAT LOOP (Used by both Manager and Worker)
// ----------------------------------------------------------------------

async function chatLoop(messages, model, signal, num_ctx, tools, toolExecutor, onChunk) {
    try {
        if (signal && signal.aborted) throw new Error('Aborted');

        // Prepare Payload
        const configuredModel = config.models.find(m => m.name === model);
        const defaultModelContext = configuredModel ? (configuredModel.contextLengthKB * 1024) : config.ollama.defaultContext;

        const payload = {
            model: model,
            messages: messages,
            tools: tools,
            stream: true,
            options: { num_ctx: num_ctx || defaultModelContext }
        };

        agentLogger.info(`[ChatLoop] Sending request to Ollama (Tools: ${tools.length})...`);

        const response = await axios.post(`${OLLAMA_BASE_URL}/api/chat`, payload, {
            responseType: 'stream',
            headers: getHeaders(),
            signal: signal
        });

        let fullContent = '';
        let toolCalls = [];
        let stats = null;

        for await (const chunk of response.data) {
            if (signal && signal.aborted) {
                response.data.destroy();
                throw new Error('Aborted');
            }

            const lines = chunk.toString().split('\n').filter(line => line.trim() !== '');
            for (const line of lines) {
                try {
                    const json = JSON.parse(line);

                    if (json.done) {
                        stats = {
                            total_duration: json.total_duration,
                            load_duration: json.load_duration,
                            prompt_eval_count: json.prompt_eval_count,
                            eval_count: json.eval_count
                        };
                    }

                    if (json.message) {
                        const content = json.message.content || '';
                        fullContent += content;

                        if (content && onChunk) {
                            onChunk({ type: 'token', content: content });
                        }

                        if (json.message.tool_calls) {
                            json.message.tool_calls.forEach(tc => toolCalls.push(tc));
                        }
                    }
                } catch (e) { /* ignore */ }
            }
        }

        const message = { role: 'assistant', content: fullContent };
        if (stats) message.stats = stats;
        if (toolCalls.length > 0) message.tool_calls = toolCalls;

        if (message.tool_calls && message.tool_calls.length > 0) {
            agentLogger.info(`Tool calls: ${message.tool_calls.length}`);
            const nextMessages = [...messages, message];

            for (const toolCall of message.tool_calls) {
                if (signal && signal.aborted) throw new Error('Aborted');
                const functionName = toolCall.function.name;
                const args = toolCall.function.arguments;

                // Execute via the provided executor (Manager or Worker logic)
                const toolResult = await toolExecutor(functionName, args);

                nextMessages.push({
                    role: 'tool',
                    content: toolResult,
                    name: functionName,
                });
            }
            // Recursion
            return chatLoop(nextMessages, model, signal, num_ctx, tools, toolExecutor, onChunk);
        }

        return message;

    } catch (e) {
        agentLogger.error(`ChatLoop Error Stack: ${e.stack}`);
        throw e;
    }
}

async function listModels() {
    return config.models.map(model => ({
        name: model.name,
        context_length: model.contextLengthKB * 1024
    }));
}

module.exports = {
    chatWithOllama: processUserRequest,
    listModels
};

File: server_py/services/ollama.py
import httpx
import json
import logging
import asyncio
from typing import List, Optional, Callable
from config import settings
from services.prompts import MANAGER_PROMPT, WORKER_PROMPT, DEEP_RESEARCH_PROMPT
from services.tools import manager_tools, worker_tools, execute_worker_tool, LEX_API_URL
from services.web_search import search_web
from services.learning import get_relevant_examples

logger = logging.getLogger("lexchat.ollama")

# ----------------------------------------------------------------------
# SHARED CHAT LOOP (ReAct)
# ----------------------------------------------------------------------

async def chat_loop(messages: list, model: str, signal_check: Callable, num_ctx: int, tools: list, tool_executor: Callable, on_chunk: Callable):
    """
    Generic ReAct loop for both Manager and Worker agents.
    """
    try:
        if signal_check and signal_check():
            raise Exception("Aborted")

        # Payload
        # Note: Python config models list might need to clearly map to context length if dynamic.
        # For now using simple fallback.
        payload = {
            "model": model,
            "messages": messages,
            "tools": tools,
            "stream": True,
            "options": {"num_ctx": num_ctx}
        }
        
        logger.info(f"[ChatLoop] Sending request to Ollama (Tools: {len(tools)})...")
        
        url = f"{settings.OLLAMA_BASE_URL}/api/chat"
        full_content = ""
        tool_calls = []
        
        async with httpx.AsyncClient(timeout=300.0) as client:
             async with client.stream("POST", url, json=payload) as response:
                if response.status_code != 200:
                    err = await response.aread()
                    logger.error(f"Ollama Error {response.status_code}: {err}")
                    raise Exception(f"Ollama Error: {response.status_code}")
                    
                async for chunk in response.aiter_bytes():
                     if signal_check and signal_check():
                         raise Exception("Aborted")
                     
                     try:
                         obj = json.loads(chunk)
                         if "message" in obj:
                             msg = obj["message"]
                             content = msg.get("content", "")
                             full_content += content
                             
                             if content and on_chunk:
                                 await on_chunk({"type": "token", "content": content})
                                 
                             if "tool_calls" in msg:
                                 tool_calls.extend(msg["tool_calls"])
                                 
                     except json.JSONDecodeError:
                         continue

        # Process Tool Calls
        message = {"role": "assistant", "content": full_content}
        if tool_calls:
            message["tool_calls"] = tool_calls
            
        if tool_calls:
            logger.info(f"Tool calls: {len(tool_calls)}")
            next_messages = messages + [message]
            
            for tc in tool_calls:
                if signal_check and signal_check():
                    raise Exception("Aborted")
                    
                func_name = tc["function"]["name"]
                args = tc["function"]["arguments"]
                
                # Execute
                tool_result = await tool_executor(func_name, args)
                
                next_messages.append({
                    "role": "tool",
                    "content": tool_result,
                    "name": func_name
                })
                
            # Recursion
            return await chat_loop(next_messages, model, signal_check, num_ctx, tools, tool_executor, on_chunk)
            
        return full_content

    except Exception as e:
        logger.error(f"ChatLoop Error: {e}")
        raise e

# ----------------------------------------------------------------------
# WORKER AGENT
# ----------------------------------------------------------------------

async def run_worker_agent(query: str, model: str, signal_check: Callable, num_ctx: int, parent_on_chunk: Callable):
    logger.info(f"[Worker] Starting research on: {query}")
    
    messages = [
        {"role": "system", "content": WORKER_PROMPT},
        {"role": "user", "content": query}
    ]
    
    # Worker tool executor wrapper for updates
    async def worker_tool_executor_wrapper(name, args):
        if parent_on_chunk:
            await parent_on_chunk({"type": "tool_start", "tool": f"Worker: {name}"})
        result = await execute_worker_tool(name, args)
        if parent_on_chunk:
            await parent_on_chunk({"type": "tool_end", "tool": f"Worker: {name}", "result": "Done"})
        return result

    # Pass None for on_chunk to suppress worker token streaming to client (Manager aggregates it)
    return await chat_loop(messages, model, signal_check, num_ctx, worker_tools, worker_tool_executor_wrapper, None)

# ----------------------------------------------------------------------
# MANAGER AGENT
# ----------------------------------------------------------------------

async def process_user_request(messages: list, model: str, on_chunk: Callable, signal_check: Callable, num_ctx: int):
    # 1. Prepare System Prompt
    system_message = {"role": "system", "content": MANAGER_PROMPT}
    
    # 2. In-Context Learning Injection
    try:
        last_msg = messages[-1] if messages else None
        if last_msg and last_msg.get("role") == "user":
            learning_data = await get_relevant_examples(last_msg.get("content"))
            # Format learning context (Simple version inline)
            context_inj = ""
            if learning_data["critiques"]:
                context_inj += "\\n### CRITICAL FEEDBACK\\nAVOID these mistakes:\\n"
                for c in learning_data["critiques"]:
                    context_inj += f"- {c.get('feedback_comment')}\\n"
            if learning_data["examples"]:
                context_inj += "\\n### SUCCESSFUL EXAMPLES\\n"
                for ex in learning_data["examples"]:
                    context_inj += f"Q: {ex.get('question')}\\nA: {ex.get('answer')}\\n---\\n"
            
            if context_inj:
                logger.info("[Learning] Injecting feedback context.")
                system_message["content"] += f"\\n\\n{context_inj}"
    except Exception as e:
        logger.error(f"[Learning] Failed to inject context: {e}")

    # Prepend System Prompt
    # Python dict copy
    final_messages = [system_message] + [m for m in messages if m.get("role") != "system"] # simple dedupe of system

    # 3. Manager Tool Executor
    async def manager_tool_executor(name, args):
        if name == "delegate_research":
            if on_chunk:
                await on_chunk({"type": "tool_start", "tool": "Research Agent"})
                
            result = await run_worker_agent(args.get("query"), model, signal_check, num_ctx, on_chunk)
            
            if on_chunk:
                await on_chunk({"type": "tool_end", "tool": "Research Agent", "result": "Research Complete"})
                
            return f"[Research Agent Result]\\n{result}"
        return f"Error: Unknown manager tool {name}"

    return await chat_loop(final_messages, model, signal_check, num_ctx, manager_tools, manager_tool_executor, on_chunk)

# ----------------------------------------------------------------------
# DEEP RESEARCH AGENT
# ----------------------------------------------------------------------

async def chat_with_deep_research(messages: list, model: str, on_status_update: Callable, signal_check: Callable, num_ctx: int):
    logger.info("[Deep Research] Starting session...")
    
    system_message = {"role": "system", "content": DEEP_RESEARCH_PROMPT}
    final_messages = [system_message] + [m for m in messages if m.get("role") != "system"]

    # Combine tools: Worker Tools + Web Search
    web_tool = {
        "type": "function",
        "function": {
            "name": "search_web",
            "description": "Search the public web for information. Use this for broad context, news, or general knowledge not in the legal database.",
            "parameters": {
                "type": "object",
                "properties": {
                    "query": {"type": "string", "description": "The search query."}
                },
                "required": ["query"]
            }
        }
    }
    deep_tools = worker_tools + [web_tool]

    async def deep_tool_executor(name, args):
        if on_status_update:
            await on_status_update({"type": "tool_start", "tool": f"Deep Research: {name}"})
            
        if name == "search_web":
            result = await search_web(args.get("query"))
        else:
            result = await execute_worker_tool(name, args)
            
        if on_status_update:
            await on_status_update({"type": "tool_end", "tool": f"Deep Research: {name}", "result": "Done"})
        return result

    return await chat_loop(final_messages, model, signal_check, num_ctx, deep_tools, deep_tool_executor, on_status_update)

# Aliases for export
chat_with_ollama = process_user_request
